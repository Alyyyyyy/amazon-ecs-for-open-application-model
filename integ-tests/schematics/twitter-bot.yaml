apiVersion: core.oam.dev/v1alpha1
kind: ApplicationConfiguration
metadata:
  name: twitter-bot
  annotations:
    version: v1.0.0
    description: "Example Twitter bot app"
spec:
  components:
    - componentName: frontend
      instanceName: web-front-end
      parameterValues:
        - name: username
          value: hello
        - name: password
          value: world
        - name: backend-address
          value: http://hello.world

    - componentName: admin-backend
      instanceName: backend-svc
      parameterValues:
        - name: twitter-consumer-key
          value: key
        - name: twitter-consumer-secret
          value: secret
        - name: twitter-access-token
          value: token
        - name: twitter-access-token-secret
          value: token-secret
---
apiVersion: core.oam.dev/v1alpha1
kind: ComponentSchematic
metadata:
  name: frontend
  annotations:
    version: v1.0.0
    description: >
      Sample component schematic that describes the administrative interface for our Twitter bot.
spec:
  workloadType: core.oam.dev/v1alpha1.Server
  osType: linux
  parameters:
  - name: username
    description: Basic auth username for accessing the administrative interface
    type: string
    required: true
  - name: password
    description: Basic auth password for accessing the administrative interface
    type: string
    required: true
  - name: backend-address
    description: Host name or IP of the backend
    type: string
    required: true
  containers:
  - name: my-twitter-bot-frontend
    image: example/my-twitter-bot-frontend@sha256:6c3c624b58dbbcd3c0dd82b4c53f04194d1247c6eebdaab7c610cf7d66709b3b
    resources:
      cpu:
        required: 2.0
      memory:
        required: 1G
    ports:
    - name: http
      containerPort: 8080
    env:
    - name: USERNAME
      fromParam: 'username'
    - name: PASSWORD
      fromParam: 'password'
    - name: BACKEND_ADDRESS
      fromParam: 'backend-address'
    livenessProbe:
      httpGet:
        port: 8080
        path: /healthz
    readinessProbe:
      httpGet:
        port: 8080
        path: /healthz
---
apiVersion: core.oam.dev/v1alpha1
kind: ComponentSchematic
metadata:
  name: admin-backend
  annotations:
    version: v1.0.0
    description: >
      Sample component schematic that describes the backend for our Twitter bot.
spec:
  workloadType: core.oam.dev/v1alpha1.Server
  osType: linux
  parameters:
  - name: twitter-consumer-key
    description: Twitter API consumer key
    type: string
    required: true
  - name: twitter-consumer-secret
    description: Twitter API consumer secret
    type: string
    required: true
  - name: twitter-access-token
    description: Twitter API access token
    type: string
    required: true
  - name: twitter-access-token-secret
    description: Twitter API access token secret
    type: string
    required: true
  containers:
  - name: my-twitter-bot-backend
    image: example/my-twitter-bot-backend@sha256:6c3c624b58dbbcd3c0dd82b4c53f04194d1247c6eebdaab7c610cf7d66709b3b
    resources:
      cpu:
        required: 1.0
      memory:
        required: 100M
      volumes:
      - name: config
        mountPath: /var/lib/my-twitter-bot/conf
        accessMode: RW
        sharingPolicy: Exclusive
    ports:
    - name: http
      containerPort: 8080
    env:
    - name: TWITTER_CONSUMER_KEY
      fromParam: 'twitter-consumer-key'
    - name: TWITTER_CONSUMER_SECRET
      fromParam: 'twitter-consumer-secret'
    - name: TWITTER_ACCESS_TOKEN
      fromParam: 'twitter-access-token'
    - name: TWITTER_ACCESS_TOKEN_SECRET
      fromParam: 'twitter-access-token-secret'
    livenessProbe:
      httpGet:
        port: 8080
        path: /healthz
    readinessProbe:
      httpGet:
        port: 8080
        path: /healthz