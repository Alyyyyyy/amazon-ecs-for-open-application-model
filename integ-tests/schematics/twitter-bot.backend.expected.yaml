AWSTemplateFormatVersion: 2010-09-09
Description: Amazon ECS infrastructure for twitter-bot backend-svc

Resources:
  LogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      LogGroupName: oam-ecs-twitter-bot-backend-svc

  TaskDefinition:
    Type: AWS::ECS::TaskDefinition
    Properties:
      Family: oam-ecs-twitter-bot-backend-svc
      NetworkMode: awsvpc
      RequiresCompatibilities:
        - FARGATE
      Cpu: 1.00 vcpu
      Memory: '2048'
      ExecutionRoleArn: !GetAtt ExecutionRole.Arn
      ContainerDefinitions:
        - Name: my-twitter-bot-backend
          Image: example/my-twitter-bot-backend@sha256:6c3c624b58dbbcd3c0dd82b4c53f04194d1247c6eebdaab7c610cf7d66709b3b
          MountPoints:
            - ContainerPath: /var/lib/my-twitter-bot/conf
              ReadOnly:  false
              SourceVolume: config
          Environment:
            - Name: TWITTER_CONSUMER_KEY
              Value:  "key"
            - Name: TWITTER_CONSUMER_SECRET
              Value:  "secret"
            - Name: TWITTER_ACCESS_TOKEN
              Value:  "token"
            - Name: TWITTER_ACCESS_TOKEN_SECRET
              Value:  "token-secret"
          PortMappings:
            - ContainerPort: 8080
              Protocol:  tcp
          LogConfiguration:
            LogDriver: awslogs
            Options:
              awslogs-region: !Ref AWS::Region
              awslogs-group: !Ref LogGroup
              awslogs-stream-prefix: oam-ecs
      Volumes:
        - Name: config

  ExecutionRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Effect: Allow
            Principal:
              Service: ecs-tasks.amazonaws.com
            Action: 'sts:AssumeRole'

      ManagedPolicyArns:
        - !Sub 'arn:${AWS::Partition}:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy'

  ContainerSecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: oam-ecs-twitter-bot-backend-svc-ContainerSecurityGroup
      VpcId:
        Fn::ImportValue: oam-ecs-VpcId

  Service:
    Type: AWS::ECS::Service
    Properties:
      Cluster:
        Fn::ImportValue: oam-ecs-ECSCluster
      TaskDefinition: !Ref TaskDefinition
      DeploymentConfiguration:
        MinimumHealthyPercent: 100
        MaximumPercent: 200
      DesiredCount: 1
      LaunchType: FARGATE
      NetworkConfiguration:
        AwsvpcConfiguration:
          AssignPublicIp: DISABLED
          Subnets:
            Fn::Split:
              - ','
              - Fn::ImportValue: oam-ecs-PrivateSubnets
          SecurityGroups:
            - !Ref ContainerSecurityGroup
      LoadBalancers:
        - ContainerName: my-twitter-bot-backend
          ContainerPort: 8080
          TargetGroupArn: !Ref TargetGroupMyTwitterBotBackend8080
      HealthCheckGracePeriodSeconds: 0
    DependsOn:
      - LBListenerMyTwitterBotBackend8080



  SGLoadBalancerToContainers:
    Type: AWS::EC2::SecurityGroupIngress
    Properties:
      Description: Ingress from anywhere on the internet through the public NLB
      GroupId: !Ref ContainerSecurityGroup
      IpProtocol: '-1'
      CidrIp: 0.0.0.0/0

  PublicLoadBalancer:
    Type: AWS::ElasticLoadBalancingV2::LoadBalancer
    Properties:
      Type: network
      Scheme: internet-facing
      Subnets:
        Fn::Split:
          - ','
          - Fn::ImportValue: oam-ecs-PublicSubnets

  LBListenerMyTwitterBotBackend8080:
    Type: AWS::ElasticLoadBalancingV2::Listener
    Properties:
      DefaultActions:
        - TargetGroupArn: !Ref TargetGroupMyTwitterBotBackend8080
          Type: 'forward'
      LoadBalancerArn: !Ref 'PublicLoadBalancer'
      Port: 8080
      Protocol:  TCP

  TargetGroupMyTwitterBotBackend8080:
    Type: AWS::ElasticLoadBalancingV2::TargetGroup
    Properties:
      Protocol:  TCP
      TargetType: ip
      Port: 8080
      VpcId:
        Fn::ImportValue: oam-ecs-VpcId
      TargetGroupAttributes:
      - Key: deregistration_delay.timeout_seconds
        Value: '30'

      HealthCheckProtocol: HTTP
      HealthCheckPath: /healthz
      HealthCheckPort: '8080'
      HealthCheckTimeoutSeconds:  6

      HealthCheckIntervalSeconds:  10
      HealthyThresholdCount:  2
      UnhealthyThresholdCount:  3



Outputs:
  CloudFormationStackConsole:
    Description: The AWS console deep-link for the CloudFormation stack
    Value: !Sub https://console.aws.amazon.com/cloudformation/home?region=${AWS::Region}#/stacks/stackinfo?stackId=${AWS::StackName}

  ECSServiceConsole:
    Description: The AWS console deep-link for the ECS service
    Value: !Sub https://console.aws.amazon.com/ecs/home?region=${AWS::Region}#/clusters/oam-ecs/services/${Service.Name}

  MyTwitterBotBackendPort8080Endpoint:
    Description: The endpoint for container MyTwitterBotBackend on port 8080
    Value: !Sub '${PublicLoadBalancer.DNSName}:8080'

